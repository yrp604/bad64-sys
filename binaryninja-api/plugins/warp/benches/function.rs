use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use criterion::{criterion_group, criterion_main, Criterion};
use rayon::prelude::*;
use std::path::PathBuf;
use warp_ninja::build_function;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 1] = ["atox.obj"];

pub fn function_benchmark(c: &mut Criterion) {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");

        let functions: Vec<_> = view.functions().iter().map(|f| f.to_owned()).collect();
        // Bench all functions sequentially
        c.bench_function("signature all functions", |b| {
            b.iter(|| {
                for func in &functions {
                    let _ = build_function(&func, || func.lifted_il().ok(), false);
                }
            })
        });

        // Bench all functions in parallel
        c.bench_function("signature all functions rayon", |b| {
            b.iter(|| {
                functions
                    .par_iter()
                    .filter_map(|func| {
                        Some(build_function(
                            func.as_ref(),
                            || func.lifted_il().ok(),
                            false,
                        ))
                    })
                    .collect::<Vec<_>>()
            })
        });
    }
}

criterion_group!(benches, function_benchmark);
criterion_main!(benches);
