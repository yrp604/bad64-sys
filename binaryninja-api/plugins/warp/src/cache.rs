use crate::convert::{from_bn_symbol, from_bn_type_internal};
use crate::{build_function, function_guid};
use binaryninja::architecture::Architecture;
use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::confidence::MAX_CONFIDENCE;
use binaryninja::function::Function as BNFunction;
use binaryninja::low_level_il::function::{
    FunctionMutability, LowLevelILFunction, NonSSA, RegularNonSSA,
};
use binaryninja::low_level_il::RegularLowLevelILFunction;
use binaryninja::rc::Guard;
use binaryninja::rc::Ref as BNRef;
use binaryninja::symbol::Symbol as BNSymbol;
use binaryninja::types::NamedTypeReference as BNNamedTypeReference;
use binaryninja::ObjectDestructor;
use dashmap::mapref::one::Ref;
use dashmap::DashMap;
use std::collections::HashSet;
use std::hash::{DefaultHasher, Hash, Hasher};
use std::sync::OnceLock;
use warp::r#type::ComputedType;
use warp::signature::function::constraints::FunctionConstraint;
use warp::signature::function::{Function, FunctionGUID};

pub static MATCHED_FUNCTION_CACHE: OnceLock<DashMap<ViewID, MatchedFunctionCache>> =
    OnceLock::new();
pub static FUNCTION_CACHE: OnceLock<DashMap<ViewID, FunctionCache>> = OnceLock::new();
pub static GUID_CACHE: OnceLock<DashMap<ViewID, GUIDCache>> = OnceLock::new();
pub static TYPE_REF_CACHE: OnceLock<DashMap<ViewID, TypeRefCache>> = OnceLock::new();

pub fn register_cache_destructor() {
    pub static mut CACHE_DESTRUCTOR: CacheDestructor = CacheDestructor;
    #[allow(static_mut_refs)]
    // SAFETY: This can be done as the backing data is an opaque ZST.
    unsafe {
        CACHE_DESTRUCTOR.register()
    };
}

pub fn cached_function_match<F>(function: &BNFunction, f: F) -> Option<Function>
where
    F: Fn() -> Option<Function>,
{
    let view = function.view();
    let view_id = ViewID::from(view.as_ref());
    let function_id = FunctionID::from(function);
    let function_cache = MATCHED_FUNCTION_CACHE.get_or_init(Default::default);
    match function_cache.get(&view_id) {
        Some(cache) => cache.get_or_insert(&function_id, f).to_owned(),
        None => {
            let cache = MatchedFunctionCache::default();
            let matched = cache.get_or_insert(&function_id, f).to_owned();
            function_cache.insert(view_id, cache);
            matched
        }
    }
}

pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let view = function.view();
    let view_id = ViewID::from(view);
    let function_id = FunctionID::from(function);
    let function_cache = MATCHED_FUNCTION_CACHE.get_or_init(Default::default);
    function_cache
        .get(&view_id)?
        .get(&function_id)?
        .value()
        .to_owned()
}

pub fn cached_function<A: Architecture>(
    function: &BNFunction,
    llil: &RegularLowLevelILFunction<A>,
) -> Function {
    let view = function.view();
    let view_id = ViewID::from(view.as_ref());
    let function_cache = FUNCTION_CACHE.get_or_init(Default::default);
    match function_cache.get(&view_id) {
        Some(cache) => cache.function(function, llil),
        None => {
            let cache = FunctionCache::default();
            let function = cache.function(function, llil);
            function_cache.insert(view_id, cache);
            function
        }
    }
}

pub fn cached_call_site_constraints(function: &BNFunction) -> HashSet<FunctionConstraint> {
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    match guid_cache.get(&view_id) {
        Some(cache) => cache.call_site_constraints(function),
        None => {
            let cache = GUIDCache::default();
            let constraints = cache.call_site_constraints(function);
            guid_cache.insert(view_id, cache);
            constraints
        }
    }
}

pub fn cached_adjacency_constraints<F>(
    function: &BNFunction,
    filter: F,
) -> HashSet<FunctionConstraint>
where
    F: Fn(&BNFunction) -> bool,
{
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    match guid_cache.get(&view_id) {
        Some(cache) => cache.adjacency_constraints(function, filter),
        None => {
            let cache = GUIDCache::default();
            let constraints = cache.adjacency_constraints(function, filter);
            guid_cache.insert(view_id, cache);
            constraints
        }
    }
}

pub fn cached_function_guid<A: Architecture, M: FunctionMutability>(
    function: &BNFunction,
    llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
) -> FunctionGUID {
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    match guid_cache.get(&view_id) {
        Some(cache) => cache.function_guid(function, llil),
        None => {
            let cache = GUIDCache::default();
            let guid = cache.function_guid(function, llil);
            guid_cache.insert(view_id, cache);
            guid
        }
    }
}

pub fn try_cached_function_guid(function: &BNFunction) -> Option<FunctionGUID> {
    let view = function.view();
    let view_id = ViewID::from(view);
    let guid_cache = GUID_CACHE.get_or_init(Default::default);
    guid_cache.get(&view_id)?.try_function_guid(function)
}

pub fn cached_type_reference(
    view: &BinaryView,
    visited_refs: &mut HashSet<TypeRefID>,
    type_ref: &BNNamedTypeReference,
) -> Option<ComputedType> {
    let view_id = ViewID::from(view);
    let type_ref_cache = TYPE_REF_CACHE.get_or_init(Default::default);
    match type_ref_cache.get(&view_id) {
        Some(cache) => cache.cached_type_reference(view, visited_refs, type_ref),
        None => {
            let cache = TypeRefCache::default();
            let ntr = cache.cached_type_reference(view, visited_refs, type_ref);
            type_ref_cache.insert(view_id, cache);
            ntr
        }
    }
}

pub fn cached_type_references(view: &BinaryView) -> Option<Ref<ViewID, TypeRefCache>> {
    let view_id = ViewID::from(view);
    let type_ref_cache = TYPE_REF_CACHE.get_or_init(Default::default);
    type_ref_cache.get(&view_id)
}

#[derive(Clone, Debug, Default)]
pub struct MatchedFunctionCache {
    pub cache: DashMap<FunctionID, Option<Function>>,
}

impl MatchedFunctionCache {
    pub fn get_or_insert<F>(
        &self,
        function_id: &FunctionID,
        f: F,
    ) -> Ref<'_, FunctionID, Option<Function>>
    where
        F: FnOnce() -> Option<Function>,
    {
        self.cache.get(function_id).unwrap_or_else(|| {
            self.cache.insert(*function_id, f());
            self.cache.get(function_id).unwrap()
        })
    }

    pub fn get(&self, function_id: &FunctionID) -> Option<Ref<'_, FunctionID, Option<Function>>> {
        self.cache.get(function_id)
    }
}

#[derive(Clone, Debug, Default)]
pub struct FunctionCache {
    pub cache: DashMap<FunctionID, Function>,
}

impl FunctionCache {
    pub fn function<A: Architecture>(
        &self,
        function: &BNFunction,
        llil: &RegularLowLevelILFunction<A>,
    ) -> Function {
        let function_id = FunctionID::from(function);
        match self.cache.get(&function_id) {
            Some(function) => function.value().to_owned(),
            None => {
                let function = build_function(function, llil);
                self.cache.insert(function_id, function.clone());
                function
            }
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct GUIDCache {
    pub cache: DashMap<FunctionID, FunctionGUID>,
}

impl GUIDCache {
    pub fn call_site_constraints(&self, function: &BNFunction) -> HashSet<FunctionConstraint> {
        let view = function.view();
        let func_id = FunctionID::from(function);
        let func_start = function.start();
        let func_platform = function.platform();
        let mut constraints = HashSet::new();
        for call_site in &function.call_sites() {
            for cs_ref_addr in view.code_refs_from_addr(call_site.address, Some(function)) {
                match view.function_at(&func_platform, cs_ref_addr) {
                    Some(cs_ref_func) => {
                        // Call site is a function, constrain on it.
                        let cs_ref_func_id = FunctionID::from(cs_ref_func.as_ref());
                        if cs_ref_func_id != func_id {
                            let call_site_offset: i64 =
                                call_site.address.wrapping_sub(func_start) as i64;
                            // TODO: If the function is thunk we should also insert the called function.
                            constraints
                                .insert(self.function_constraint(&cs_ref_func, call_site_offset));
                        }
                    }
                    None => {
                        // We could be dealing with an extern symbol, get the symbol as a constraint.
                        let call_site_offset: i64 =
                            call_site.address.wrapping_sub(func_start) as i64;
                        if let Some(call_site_sym) = view.symbol_by_address(cs_ref_addr) {
                            constraints.insert(
                                self.function_constraint_from_symbol(
                                    &call_site_sym,
                                    call_site_offset,
                                ),
                            );
                        }
                    }
                }
            }
        }
        constraints
    }

    pub fn adjacency_constraints<F>(
        &self,
        function: &BNFunction,
        filter: F,
    ) -> HashSet<FunctionConstraint>
    where
        F: Fn(&BNFunction) -> bool,
    {
        let view = function.view();
        let func_id = FunctionID::from(function);
        let func_start = function.start();
        let mut constraints = HashSet::new();

        let mut func_addr_constraint = |func_start_addr| {
            // NOTE: We could potentially have dozens of functions all at the same start address.
            for curr_func in &view.functions_at(func_start_addr) {
                let curr_func_id = FunctionID::from(curr_func.as_ref());
                if curr_func_id != func_id && filter(curr_func.as_ref()) {
                    // NOTE: For this to work the GUID has to have already been cached. If not it will just be the symbol.
                    // Function adjacent to another function, constrain on the pattern.
                    let curr_addr_offset = (func_start_addr as i64) - func_start as i64;
                    constraints.insert(self.function_constraint(&curr_func, curr_addr_offset));
                }
            }
        };

        let mut before_func_start = func_start;
        for _ in 0..2 {
            before_func_start = view.function_start_before(before_func_start);
            func_addr_constraint(before_func_start);
        }

        let mut after_func_start = func_start;
        for _ in 0..2 {
            after_func_start = view.function_start_after(after_func_start);
            func_addr_constraint(after_func_start);
        }

        constraints
    }

    /// Construct a function constraint, must pass the offset at which it is located.
    pub fn function_constraint(&self, function: &BNFunction, offset: i64) -> FunctionConstraint {
        let guid = self.try_function_guid(function);
        let symbol = from_bn_symbol(&function.symbol());
        FunctionConstraint {
            guid,
            symbol: Some(symbol),
            offset,
        }
    }

    /// Construct a function constraint from a symbol, typically used for extern function call sites, must pass the offset at which it is located.
    pub fn function_constraint_from_symbol(
        &self,
        symbol: &BNSymbol,
        offset: i64,
    ) -> FunctionConstraint {
        let symbol = from_bn_symbol(symbol);
        FunctionConstraint {
            guid: None,
            symbol: Some(symbol),
            offset,
        }
    }

    pub fn function_guid<A: Architecture, M: FunctionMutability>(
        &self,
        function: &BNFunction,
        llil: &LowLevelILFunction<A, M, NonSSA<RegularNonSSA>>,
    ) -> FunctionGUID {
        let function_id = FunctionID::from(function);
        match self.cache.get(&function_id) {
            Some(function_guid) => function_guid.value().to_owned(),
            None => {
                let function_guid = function_guid(function, llil);
                self.cache.insert(function_id, function_guid);
                function_guid
            }
        }
    }

    pub fn try_function_guid(&self, function: &BNFunction) -> Option<FunctionGUID> {
        let function_id = FunctionID::from(function);
        self.cache
            .get(&function_id)
            .map(|function_guid| function_guid.value().to_owned())
    }
}

#[derive(Clone, Debug, Default)]
pub struct TypeRefCache {
    pub cache: DashMap<TypeRefID, Option<ComputedType>>,
}

impl TypeRefCache {
    /// NOTE: No self-referential type must be used on this function.
    pub fn cached_type_reference(
        &self,
        view: &BinaryView,
        visited_refs: &mut HashSet<TypeRefID>,
        type_ref: &BNNamedTypeReference,
    ) -> Option<ComputedType> {
        let ntr_id = TypeRefID::from(type_ref);
        match self.cache.get(&ntr_id) {
            Some(cache) => cache.to_owned(),
            None => match type_ref.target(view) {
                Some(raw_ty) => {
                    let computed_ty = ComputedType::new(from_bn_type_internal(
                        view,
                        visited_refs,
                        &raw_ty,
                        MAX_CONFIDENCE,
                    ));
                    self.cache
                        .entry(ntr_id)
                        .insert(Some(computed_ty))
                        .to_owned()
                }
                None => self.cache.entry(ntr_id).insert(None).to_owned(),
            },
        }
    }
}

/// A unique view ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct ViewID(u64);

impl From<&BinaryView> for ViewID {
    fn from(value: &BinaryView) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.original_image_base());
        hasher.write_usize(value.file().session_id());
        Self(hasher.finish())
    }
}

impl From<BNRef<BinaryView>> for ViewID {
    fn from(value: BNRef<BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BinaryView>> for ViewID {
    fn from(value: Guard<'_, BinaryView>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique function ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct FunctionID(u64);

impl From<&BNFunction> for FunctionID {
    fn from(value: &BNFunction) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write_u64(value.start());
        hasher.write_u64(value.lowest_address());
        hasher.write_u64(value.highest_address());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNFunction>> for FunctionID {
    fn from(value: BNRef<BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNFunction>> for FunctionID {
    fn from(value: Guard<'_, BNFunction>) -> Self {
        Self::from(value.as_ref())
    }
}

/// A unique named type reference ID, used for caching.
#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct TypeRefID(u64);

impl From<&BNNamedTypeReference> for TypeRefID {
    fn from(value: &BNNamedTypeReference) -> Self {
        let mut hasher = DefaultHasher::new();
        hasher.write(value.id().as_bytes());
        Self(hasher.finish())
    }
}

impl From<BNRef<BNNamedTypeReference>> for TypeRefID {
    fn from(value: BNRef<BNNamedTypeReference>) -> Self {
        Self::from(value.as_ref())
    }
}

impl From<Guard<'_, BNNamedTypeReference>> for TypeRefID {
    fn from(value: Guard<'_, BNNamedTypeReference>) -> Self {
        Self::from(value.as_ref())
    }
}

pub struct CacheDestructor;

impl ObjectDestructor for CacheDestructor {
    fn destruct_view(&self, view: &BinaryView) {
        // Clear caches as the view is no longer alive.
        let view_id = ViewID::from(view);
        if let Some(cache) = MATCHED_FUNCTION_CACHE.get() {
            cache.remove(&view_id);
        }
        if let Some(cache) = FUNCTION_CACHE.get() {
            cache.remove(&view_id);
        }
        if let Some(cache) = GUID_CACHE.get() {
            cache.remove(&view_id);
        }
        if let Some(cache) = TYPE_REF_CACHE.get() {
            cache.remove(&view_id);
        }
        log::debug!("Removed WARP caches for {:?}", view.file().filename());
    }
}
