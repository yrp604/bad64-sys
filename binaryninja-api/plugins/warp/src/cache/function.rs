use crate::convert::{comment_to_bn_comment, to_bn_symbol_at_address};
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::function::{Function as BNFunction, FunctionUpdateType};
use binaryninja::symbol::SymbolType;
use warp::signature::function::Function;

// TODO: Rename this?
/// Inserts a function match into the cache. This also has the side effect of setting persisted function
/// information, such as the matched function symbol.
///
/// IMPORTANT: This will mark the function as needing updates, if you intend to fill in functions with
/// no match (i.e. `None`), then you must change this function to prevent marking that as needing updates.
/// However, it's perfectly valid to remove a match and need to update the function still, so be careful.
pub fn insert_cached_function_match(function: &BNFunction, matched_function: Option<Function>) {
    let view = function.view();
    let function_start = function.start();
    // NOTE: If we expect to run match_function multiple times on a function, we should move this elsewhere.
    // Mark the function as needing updates so that reanalysis occurs on the function, and we apply the match.
    function.mark_updates_required(FunctionUpdateType::FullAutoFunctionUpdate);
    if let Some(auto_sym) = view.symbol_by_address(function_start) {
        // TODO: If we ever create non library function symbols we will need to remove this check (see: `to_bn_symbol_at_address`).
        if auto_sym.sym_type() == SymbolType::LibraryFunction {
            // NOTE: This will also mark for full auto function update, one thing to note is that the
            // requirement to call this is that this function not be called in the associated function's analysis.
            view.undefine_auto_symbol(&auto_sym);
        }
    }
    match matched_function {
        Some(matched_function) => {
            // Define the new matched function symbol, this can safely be done here as the symbol itself
            // will be persisted, unlike function type information or variable information.
            let new_sym = to_bn_symbol_at_address(&view, &matched_function.symbol, function_start);
            view.define_auto_symbol(&new_sym);
            // TODO: How to clear the comments? They are just persisted.
            // TODO: Also they generate an undo action, i hate implicit undo actions so much.
            for comment in &matched_function.comments {
                let bn_comment = comment_to_bn_comment(&function, comment.clone());
                function.set_comment_at(bn_comment.addr, &bn_comment.comment);
            }
            function.store_metadata("warp_matched_function", &matched_function.to_bytes(), false);
        }
        None => {
            function.remove_metadata("warp_matched_function");
        }
    }
}

// TODO: This does allocations, and for every reanalysis.
pub fn try_cached_function_match(function: &BNFunction) -> Option<Function> {
    let metadata = function.query_metadata("warp_matched_function")?;
    let raw_metadata = metadata.get_raw()?;
    Function::from_bytes(&raw_metadata)
}
