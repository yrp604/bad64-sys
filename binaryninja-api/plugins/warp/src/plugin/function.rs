use crate::cache::{cached_function_guid, try_cached_function_guid};
use crate::{get_warp_include_tag_type, INCLUDE_TAG_NAME};
use binaryninja::background_task::BackgroundTask;
use binaryninja::binary_view::{BinaryView, BinaryViewExt};
use binaryninja::command::{Command, FunctionCommand};
use binaryninja::function::Function;
use binaryninja::rc::Guard;
use rayon::iter::ParallelIterator;
use std::thread;
use warp::signature::function::FunctionGUID;

pub struct IncludeFunction;

impl FunctionCommand for IncludeFunction {
    fn action(&self, view: &BinaryView, func: &Function) {
        let sym_name = func.symbol().short_name();
        let sym_name_str = sym_name.to_string_lossy();
        let should_add_tag = func.function_tags(None, Some(INCLUDE_TAG_NAME)).is_empty();
        let insert_tag_type = get_warp_include_tag_type(view);
        match should_add_tag {
            true => {
                log::info!(
                    "Including selected function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.add_tag(&insert_tag_type, "", None, false, None);
            }
            false => {
                log::info!(
                    "Removing included function '{}' at 0x{:x}",
                    sym_name_str,
                    func.start()
                );
                func.remove_tags_of_type(&insert_tag_type, None, false, None);
            }
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        // TODO: Only allow if the function is named?
        true
    }
}

pub struct CopyFunctionGUID;

impl FunctionCommand for CopyFunctionGUID {
    fn action(&self, _view: &BinaryView, func: &Function) {
        let Some(guid) = cached_function_guid(func, || func.lifted_il().ok()) else {
            log::error!("Could not get guid for copied function");
            return;
        };
        log::info!(
            "Function GUID for {:?}... {}",
            func.symbol().short_name(),
            guid
        );
        if let Ok(mut clipboard) = arboard::Clipboard::new() {
            let _ = clipboard.set_text(guid.to_string());
        }
    }

    fn valid(&self, _view: &BinaryView, _func: &Function) -> bool {
        true
    }
}

pub struct FindFunctionFromGUID;

impl Command for FindFunctionFromGUID {
    fn action(&self, view: &BinaryView) {
        let Some(guid_str) = binaryninja::interaction::get_text_line_input(
            "Function GUID",
            "Find Function from GUID",
        ) else {
            return;
        };

        let Ok(searched_guid) = guid_str.parse::<FunctionGUID>() else {
            log::error!("Failed to parse function guid... {}", guid_str);
            return;
        };

        log::info!("Searching functions for GUID... {}", searched_guid);
        let funcs = view.functions();
        let view = view.to_owned();
        thread::spawn(move || {
            let background_task = BackgroundTask::new(
                &format!("Searching functions for GUID... {}", searched_guid),
                false,
            );

            // Only run this for functions which have already generated a GUID.
            let matched: Vec<Guard<Function>> = funcs
                .par_iter()
                .filter(|func| {
                    try_cached_function_guid(func).is_some_and(|guid| guid == searched_guid)
                })
                .collect();

            if matched.is_empty() {
                log::info!("No matches found for GUID... {}", searched_guid);
            } else {
                for func in &matched {
                    // Also navigate the user, as that is probably what they want.
                    if matched.len() == 1 {
                        let current_view = view.file().current_view();
                        if view
                            .file()
                            .navigate_to(&current_view, func.start())
                            .is_err()
                        {
                            log::error!(
                                "Failed to navigate to found function 0x{:0x} in view {}",
                                func.start(),
                                current_view
                            );
                        }
                    }
                    log::info!("Match found at function... 0x{:0x}", func.start());
                }
            }

            background_task.finish();
        });
    }

    fn valid(&self, _view: &BinaryView) -> bool {
        true
    }
}
