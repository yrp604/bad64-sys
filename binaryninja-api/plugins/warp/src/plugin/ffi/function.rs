use crate::build_function;
use crate::cache::{insert_cached_function_match, try_cached_function_match};
use crate::convert::{to_bn_symbol_at_address, to_bn_type};
use crate::plugin::ffi::{BNWARPConstraint, BNWARPFunction, BNWARPFunctionGUID};
use binaryninja::function::Function;
use binaryninja::rc::Ref;
use binaryninja::string::BnString;
use binaryninjacore_sys::{BNFunction, BNSymbol, BNType};
use std::ffi::c_char;
use std::mem::ManuallyDrop;
use std::sync::Arc;
use warp::signature::comment::FunctionComment;

#[repr(C)]
pub struct BNWarpFunctionComment {
    pub text: *mut c_char,
    pub offset: i64,
}

impl BNWarpFunctionComment {
    /// Leaks the text string to be freed with BNWARPFreeFunctionComment
    pub fn from_owned(value: &FunctionComment) -> Self {
        let text = BnString::into_raw(BnString::new(&value.text));
        Self {
            text,
            offset: value.offset,
        }
    }

    pub unsafe fn free_raw(value: &Self) {
        unsafe { BnString::free_raw(value.text) }
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match build_function(&function, || function.lifted_il().ok(), false) {
        Some(function) => Arc::into_raw(Arc::new(function)) as *mut BNWARPFunction,
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPGetMatchedFunction(
    analysis_function: *mut BNFunction,
) -> *mut BNWARPFunction {
    let function = Function::from_raw(analysis_function);
    match try_cached_function_match(&function) {
        Some(matched_function) => {
            let arc_matched_function = Arc::new(matched_function);
            // NOTE: Freed by BNWARPFreeFunctionReference
            Arc::into_raw(arc_matched_function) as *mut BNWARPFunction
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionApply(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) {
    let analysis_function = Function::from_raw(analysis_function);
    match function.is_null() {
        false => {
            // Set the matched function to `function`.
            let matched_function = ManuallyDrop::new(Arc::from_raw(function));
            insert_cached_function_match(
                &analysis_function,
                Some(matched_function.as_ref().clone()),
            )
        }
        true => {
            // We are removing the previous match.
            insert_cached_function_match(&analysis_function, None)
        }
    };
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetGUID(
    function: *mut BNWARPFunction,
) -> BNWARPFunctionGUID {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    function.guid
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbol(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNSymbol {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let view = analysis_function.view();
    let address = analysis_function.symbol().address();
    let function_symbol = to_bn_symbol_at_address(&view, &function.symbol, address);
    // NOTE: The symbol ref has been pre-incremented for the caller.
    Ref::into_raw(function_symbol).handle
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetSymbolName(function: *mut BNWARPFunction) -> *mut c_char {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let bn_name = BnString::new(&function.symbol.name);
    // NOTE: The symbol name string to be freed by BNFreeString
    BnString::into_raw(bn_name)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetType(
    function: *mut BNWARPFunction,
    analysis_function: *mut BNFunction,
) -> *mut BNType {
    let analysis_function = Function::from_raw(analysis_function);
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    match &function.ty {
        Some(func_ty) => {
            let arch = analysis_function.arch();
            let function_type = to_bn_type(Some(arch), func_ty);
            // NOTE: The type ref has been pre-incremented for the caller.
            unsafe { Ref::into_raw(function_type) }.handle
        }
        None => std::ptr::null_mut(),
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetConstraints(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWARPConstraint {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_constraints: Box<[BNWARPConstraint]> = function
        .constraints
        .clone()
        .into_iter()
        .map(Into::into)
        .collect();
    *count = raw_constraints.len();
    let raw_constraints_ptr = Box::into_raw(raw_constraints);
    raw_constraints_ptr as *mut BNWARPConstraint
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionGetComments(
    function: *mut BNWARPFunction,
    count: *mut usize,
) -> *mut BNWarpFunctionComment {
    // We do not own function so we should not drop.
    let function = ManuallyDrop::new(Arc::from_raw(function));
    let raw_comments: Box<[_]> = function
        .comments
        .iter()
        .map(BNWarpFunctionComment::from_owned)
        .collect();
    *count = raw_comments.len();
    let raw_comments_ptr = Box::into_raw(raw_comments);
    raw_comments_ptr as *mut BNWarpFunctionComment
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFunctionsEqual(
    function_a: *mut BNWARPFunction,
    function_b: *mut BNWARPFunction,
) -> bool {
    // We do not own function so we should not drop.
    let function_a = ManuallyDrop::new(Arc::from_raw(function_a));
    // We do not own function so we should not drop.
    let function_b = ManuallyDrop::new(Arc::from_raw(function_b));
    function_a.eq(&function_b)
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPNewFunctionReference(
    function: *mut BNWARPFunction,
) -> *mut BNWARPFunction {
    Arc::increment_strong_count(function);
    function
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionReference(function: *mut BNWARPFunction) {
    if function.is_null() {
        return;
    }
    Arc::decrement_strong_count(function);
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionList(functions: *mut *mut BNWARPFunction, count: usize) {
    let functions_ptr = std::ptr::slice_from_raw_parts_mut(functions, count);
    let functions = Box::from_raw(functions_ptr);
    for function in functions {
        // NOTE: The functions themselves should also be arc.
        BNWARPFreeFunctionReference(function);
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeFunctionCommentList(
    comments: *mut BNWarpFunctionComment,
    count: usize,
) {
    let comments_ptr = std::ptr::slice_from_raw_parts_mut(comments, count);
    let comments = Box::from_raw(comments_ptr);
    for comment in &comments {
        BNWarpFunctionComment::free_raw(comment)
    }
}

#[no_mangle]
pub unsafe extern "C" fn BNWARPFreeConstraintList(
    constraints: *mut BNWARPConstraint,
    count: usize,
) {
    let constraints_ptr = std::ptr::slice_from_raw_parts_mut(constraints, count);
    let _constraints = unsafe { Box::from_raw(constraints_ptr) };
}
