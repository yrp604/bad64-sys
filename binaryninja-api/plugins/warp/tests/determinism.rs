//! This tests to make sure the function GUIDs are stable.
use binaryninja::binary_view::BinaryViewExt;
use binaryninja::headless::Session;
use std::collections::BTreeMap;
use std::path::PathBuf;
use warp::signature::function::FunctionGUID;
use warp_ninja::cache::cached_function_guid;

// These are the target files present in OUT_DIR
// Add the files to fixtures/bin
static TARGET_FILES: [&str; 10] = [
    "_ctype.obj",
    "_fptostr.obj",
    "_mbslen.obj",
    "_memicmp.obj",
    "_strnicm.obj",
    "_wctype.obj",
    "atof.obj",
    "atoldbl.obj",
    "atox.obj",
    "ls",
];

#[test]
fn insta_signatures() {
    let session = Session::new().expect("Failed to initialize session");
    let out_dir = env!("OUT_DIR").parse::<PathBuf>().unwrap();
    for file_name in TARGET_FILES {
        let path = out_dir.join(file_name);
        let view = session.load(&path).expect("Failed to load view");
        let functions: BTreeMap<u64, FunctionGUID> = view
            .functions()
            .iter()
            .map(|f| {
                let guid = cached_function_guid(&f, || f.lifted_il().ok());
                (f.start(), guid.unwrap())
            })
            .collect();
        insta::assert_debug_snapshot!(file_name, functions);
    }
}
