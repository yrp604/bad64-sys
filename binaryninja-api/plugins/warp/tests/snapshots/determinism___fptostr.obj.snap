---
source: plugins/warp/tests/determinism.rs
expression: functions
---
{
    180832: FunctionGUID {
        guid: 222305cd-0d84-5eba-8a3e-2774c9c4a5dc,
    },
}
